aries_vcx/src/common/credentials/encoding.rs

```diff
@@ -1,99 +1,157 @@
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use serde_json::Value;
+
 use crate::errors::error::{AriesVcxError, AriesVcxErrorKind, VcxResult};
 use crate::utils::openssl::encode;
-use std::collections::HashMap;
 
-pub fn encode_attributes(attributes: &str) -> VcxResult<String> {
-    let mut dictionary = HashMap::new();
-    match serde_json::from_str::<HashMap<String, serde_json::Value>>(attributes) {
-        Ok(mut attributes) => {
-            for (attr, attr_data) in attributes.iter_mut() {
-                let first_attr = match attr_data {
-                    // new style input such as {"address2":"101 Wilson Lane"}
-                    serde_json::Value::String(str_type) => str_type,
-
-                    // old style input such as {"address2":["101 Wilson Lane"]}
-                    serde_json::Value::Array(array_type) => {
-                        let attrib_value: &str = match array_type.get(0).and_then(serde_json::Value::as_str) {
-                            Some(x) => x,
-                            None => {
-                                return Err(AriesVcxError::from_msg(
-                                    AriesVcxErrorKind::InvalidAttributesStructure,
-                                    "Attribute value not found",
-                                ));
-                            }
-                        };
-
-                        warn!("Old attribute format detected. See vcx_issuer_create_credential api for additional information.");
-                        attrib_value
+/// `CredentialAttribute` contains credential attributes data in a given state.
+#[derive(Debug, Clone)]
+pub struct CredentialAttribute<'r, State> {
+    /// Raw credential attribute data.
+    raw: &'r str,
+
+    /// Encoded credential attribute data.
+    encoded: Option<String>,
```

Comment:
Let's get rid of `Option` here. It's `None` only temporarily when calling `new` function. But since it's now `Option`, we don't have quarantee it's not `None` when we return `CredentialAttribute`.

Reply:
Yes, I was not sure of the option as well because the you can only access the encoded value when you have `AttributeValue<EncodedAttributeValue>`.